lib/core/src/Cardano/Pool/Metadata.hs

```diff
@@ -70,45 +102,133 @@ defaultManagerSettings =
+    -> Either HttpException URI
+registryUrlBuilder baseUrl _ (StakePoolMetadataHash bytes) =
+    Right $ baseUrl
+        { uriPath = "/" <> intercalate "/" (pathSegments baseUrl ++ [hash])
+        }
+  where
+    hash = T.unpack $ T.decodeUtf8 $ convertToBase Base16 bytes
+
 fetchFromRemote
-    :: Manager
+    :: Tracer IO StakePoolMetadataFetchLog
+    -> [StakePoolMetadataUrl -> StakePoolMetadataHash -> Either HttpException URI]
+    -> Manager
     -> StakePoolMetadataUrl
     -> StakePoolMetadataHash
-    -> IO (Either String StakePoolMetadata)
-fetchFromRemote manager url_ hash_ = runExceptT $ do
-    req <- fromInvalidUrl $ parseUrlThrow (T.unpack url)
-    chunk <- ExceptT
-        $ handle fromIOException
-        $ handle fromHttpException
-        $ withResponse req manager $ \res -> do
-        -- NOTE
-        -- Metadata are _supposed to_ be made of:
-        --
-        -- - A name (at most 50 UTF-8 bytes)
-        -- - An optional description (at most 250 UTF-8 bytes)
-        -- - A ticker (at most 5 UTF-8 bytes)
-        -- - A homepage (at most 100 UTF-8 bytes)
-        --
-        -- So, the total, including a pretty JSON encoding with newlines ought
-        -- to be less than 512 bytes. For security reasons, we only download the
-        -- first 512 bytes.
-        pure . BL.toStrict <$> brReadSome (responseBody res) 512
-    when (blake2b256 chunk /= hash) $ throwE $ mconcat
+    -> IO (Maybe StakePoolMetadata)
+fetchFromRemote tr builders manager url hash = runExceptTLog $ do
+    chunk <- fromFirst builders
+    when (blake2b256 chunk /= coerce hash) $ throwE $ mconcat
         [ "Metadata hash mismatch. Saw: "
         , B8.unpack $ hex $ blake2b256 chunk
         , ", but expected: "
-        , B8.unpack $ hex hash
+        , B8.unpack $ hex $ coerce @_ @ByteString hash
         ]
     except $ eitherDecodeStrict chunk
   where
-    StakePoolMetadataUrl  url  = url_
-    StakePoolMetadataHash hash = hash_
-
     fromIOException :: Monad m => IOException -> m (Either String a)
     fromIOException = return . Left . ("IO exception: " <>) . show
 
-    fromHttpException :: Monad m => HttpException -> m (Either String a)
-    fromHttpException = return . Left . ("HTTP exception: " <>) . show
+    fromHttpException :: Monad m => HttpException -> m (Either String (Maybe a))
+    fromHttpException = const (return $ Right Nothing)
+
+    runExceptTLog
+        :: ExceptT String IO StakePoolMetadata
+        -> IO (Maybe StakePoolMetadata)
+    runExceptTLog action = runExceptT action >>= \case
+        Left msg ->
+            Nothing <$ traceWith tr (MsgFetchPoolMetadataFailure hash msg)
+
+        Right meta ->
+            Just meta <$ traceWith tr (MsgFetchPoolMetadataSuccess hash meta)
+
+    -- Try each builder in order, but only if the previous builder led to a
+    -- "ConnectionFailure" exception. Other exceptions mean that we could
+    -- likely reach the server.
+    fromFirst (builder:rest) = do
```

Comment:
I found this a little confusing. Can we separate the function which makes a request and reads at most 512 bytes from the response from the function which tries each builder and uses the first which succeeds?

Reply:
I can try that :+1: